pallets/loans/src/types/cashflow.rs

```diff
@@ -0,0 +1,430 @@
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_traits::{interest::InterestRate, Seconds};
+use chrono::{DateTime, Datelike, NaiveDate};
+use frame_support::pallet_prelude::RuntimeDebug;
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{
+		EnsureAdd, EnsureAddAssign, EnsureDiv, EnsureFixedPointNumber, EnsureInto, EnsureSub,
+		EnsureSubAssign,
+	},
+	ArithmeticError, DispatchError, FixedPointNumber, FixedPointOperand, FixedU128,
+};
+use sp_std::{cmp::min, vec, vec::Vec};
+
+// By now only "day 1" of the month is supported for monthly cashflows.
+// Modifying this value would make `monthly_dates()` and
+// `monthly_intervals()` to no longer work as expected.
+// Supporting more reference dates will imply more logic related to dates.
+const DAY_1: u32 = 1;
+
+/// Specify the expected repayments date
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum Maturity {
+	/// Fixed point in time, in secs
+	Fixed {
+		/// Secs when maturity ends
+		date: Seconds,
+		/// Extension in secs, without special permissions
+		extension: Seconds,
+	},
+}
+
+impl Maturity {
+	pub fn fixed(date: Seconds) -> Self {
+		Self::Fixed { date, extension: 0 }
+	}
+
+	pub fn date(&self) -> Seconds {
+		match self {
+			Maturity::Fixed { date, .. } => *date,
+		}
+	}
+
+	pub fn is_valid(&self, now: Seconds) -> bool {
+		match self {
+			Maturity::Fixed { date, .. } => *date > now,
+		}
+	}
+
+	pub fn extends(&mut self, value: Seconds) -> Result<(), ArithmeticError> {
+		match self {
+			Maturity::Fixed { date, extension } => {
+				date.ensure_add_assign(value)?;
+				extension.ensure_sub_assign(value)
+			}
+		}
+	}
+}
+
+/// Interest payment periods
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum InterestPayments {
+	/// All interest is expected to be paid at the maturity date
+	None,
+
+	/// Interest is expected to be paid monthly
+	/// The associated value correspond to the paydown day in the month,
+	/// from 1-31.
+	/// The day will be adjusted to the month.
+	///
+	/// NOTE: Only day 1 is supported by now
+	Monthly(u8),
+}
+
+/// Specify the paydown schedules of the loan
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum PayDownSchedule {
+	/// The entire borrowed amount is expected to be paid back at the maturity
+	/// date
```

Comment:
It means that no matter how you pay it (one big payment or several payments during the time), you need to repay all before the maturity. Basically, it means no restrictions on the payments. Maybe we can rewrite this to avoid misunderstandings

Reply:
[77dada6](https://github.com/centrifuge/centrifuge-chain/pull/1797/commits/77dada68c61c44f1af1ad42d5cb51d2330b4596c)